ext/pdo/pdo.stub.php

```diff
@@ -0,0 +1,128 @@
+<?php
+
+/* pdo.c */
+
+function pdo_drivers(): array {}
+
+/* pdo_dbh.c */
+
+class PDO {
+    public function __construct(string $dsn, ?string $username = null, ?string $passwd = null, ?array $options = null) {}
```

Comment:
We can't have param types as well, this will break compatibilities. See [tests failures](https://travis-ci.org/php/php-src/jobs/595859299#L662) for example.

Reply:
Hmm, are you sure? Since PHP 7.2 omitting parameter type declarations in child classes shouldn't be a problem (because of the partial covariance-contravariance RFC). Or is there any other problem with adding parameter type declarations I am not aware of? 🤔

Reply:
Nikita, could you tell me how I should proceed with the test in question? Should I adapt the IncorrectInterface according to the new interface?

Reply:
That's right.

Reply:
Unfortunately, I still don't get how to solve the issue :S

The method's stub is now the following:
```php
/** @return mixed */
public function fetch(?int $fetch_style = PDO::FETCH_BOTH, ?int $cursor_orientation = PDO::FETCH_ORI_NEXT, ?int $cursor_offset = 0) {}
```

I changed one of the declared interfaces in the test from

```php
interface StatementInterface {
	public function fetch($first = null, $second, $third);
}
```

to

```php
interface StatementInterface {
	public function fetch($first = PDO::FETCH_BOTH, $second = PDO::FETCH_ORI_NEXT, $third = 0);
}
```

And the other one looks like this:

```php
interface StatementInterface1 {
	public function fetch($first = null, $second = null, $third = null);
}
```

But I got this error message when running the test locally:

```
Fatal error: Declaration of PDOStatement::fetch(int $fetch_style = NULL, int $cursor_orientation = NULL, int $cursor_offset = NULL) must be compatible with StatementInterface::fetch($first = PDO::FETCH_BOTH, $second = PDO::FETCH_ORI_NEXT, $third = 0)
```

I don't get for example why the default params of `PDOStatement::fetch()` are all `null`? :S

Reply:
Well, the parameters are nullable, so apparently the engine assumes the default would be `null`.

Reply:
I am out of any ideas after having changed the parameters to be non-nullable (+ adapted the interface in the test), but I still get this error:
`Fatal error: Declaration of PDOStatement::fetch(int $fetch_style = NULL, int $cursor_orientation = NULL, int $cursor_offset = NULL) must be compatible with StatementInterface::fetch($first = PDO::FETCH_BOTH, $second = PDO::FETCH_ORI_NEXT, $third = 0) in /home/travis/build/php/php-src/Zend/tests/bug71428.2.php`

https://travis-ci.org/php/php-src/jobs/596831496#L661

Reply:
`StatementInterface::fetch()` does not declare any parameter types, but `PDOStatement::fetch()` does. Narrowing of parameter types is not allowed, though, since it would violate LSP. Therefore bug71428.2.phpt has to be adjusted to also declare the parameter types.

Reply:
Thanks, @cmb69, again! The `null` default parameters deceived me so much that I didn't realize that the interface should have the type declarations as well..

Reply:
@cmb69 @nikic I had a look why the default value of parameters of internal functions are reported as `null` (mainly because I was curious if a mortal soul like me could fix it :) ), and discovered/realized that the arginfo doesn't store default values (and it explains why default param values of userland functions are retrieved from opcodes). What I think is that as soon as stubs for all internal functions are generated, it should be easy to also add the default param values to arginfos - at least implementation-wise. Do my observations make any sense?

I am concerned though that the newly added default values would negatively affect performance or binary size. Are my concerns true? If not, is there any other countargument against implementing this feature? If it is worth to do, I'd love to work on it - but only if you think it's doable for someone who is new to PHP internal development (but very enthusiastic). :)

Reply:
Your analysis is correct. Internal functions don't store defaults, so NULL is used as the next best thing. Adding default value information would certainly come with a memory usage cost, because we need to store the information ... but maybe not too high.

A secondary problem is that arginfo structures for internal functions need to be compile-time constants, which are essentially impossible to construct for things like strings. It might be easiest to instead store the PHP expressions and parse them when reflection is used, rather than storing zvals. This also has the advantage of needing less space.

Reply:
@nikic Unfortunately, I don't exactly get your idea, but I also have one:

What about generating a file via the `gen_stub.php` that would contain a map of all the distinct default values found in the stubs, like below:
```C
const char *zend_parameter_default_values[] = {
    "", "true", "false", "0", "null", // ...
};
```

And then the `zend_arg_info` structure should be extended with something like
`zend_long default_value_index`, so that `default_value_index` would store the key of the appropriate value. This way, same values could be stored only once, and they could be looked up when needed.

Is this a viable approach? Or this was what you suggested? I understand this solution is not very good for PECL extensions (because they couldn't add new values to the map), but probably still better than storing same values hundreds of times.

If you have some minutes for this case, I'd be happy if you could provide a feedback. :)

Reply:
What I had in mind is just specifying `"false"` etc (as a `const char *`), which works fine in constant initializers and will be deduplicated by the linker. My comment was to use that instead of storing an explicit zval, which is hard to impossible to initialize, and also larger.

Reply:
OK, I see now, thanks for the clarification. My initial idea was the same then, but I missed the `const` qualifier when I tried it, so the string value didn't work in the initializer. That's why I tried to search for alternative solutions.

Reply:
For reference: #4832.